compiler.py

```diff
@@ -0,0 +1,637 @@
+        elif method == 'ceasefire':
+            if len(node.args) != 1:
+                raise CompilerError(ERR_BAD_SYSCALL_ARGS, node)
+
+            link = self.as_value(node.args[0])
+            self._ins.append(f'control shoot {link} 0 0 0')
+        else:
+            raise CompilerError(ERR_UNSUPPORTED_SYSCALL, node)
+
+    def as_value(self, node):
+        """This will be called by various visit methods when they encounter something to treat as an atomic expression,
+           e.g. for the y and 3 in x=y+3.  This coerces the given node to a string that can appear in mlog code."""
+
+        # TODO figure out balance of labor between this and other things in evaluating complex expressions
+
+        if isinstance(node, ast.Constant):  # constants include True, 1.5, and "yarn"
+            if isinstance(node.value, bool):
+                return ('false', 'true')[node.value]
+            elif isinstance(node.value, (int, float)):
+                return str(node.value)
+            elif isinstance(node.value, str):
+                return '"' + ''.join(c for c in node.value if c >= ' ' and c != '"') + '"' #scrub quotes and control characters from string
+            else:
+                raise CompilerError(ERR_COMPLEX_VALUE, node)
+
+        elif isinstance(node, ast.Name):
+            assert isinstance(node.ctx, ast.Load)
+            # TODO should this raise a CompilerError like most other errors do?
+            return node.id
+            # TODO figure out what this does???
+
+        elif isinstance(node, ast.Call):
+            if isinstance(node.func, ast.Attribute):
+                if node.func.value.id == 'Env':
+                    return self.env_as_value(node.func)
+                else:
+                    raise CompilerError(ERR_COMPLEX_VALUE, node)
+            # TODO  It looks like this is the place to handle object attributes like vault1.copper and Unit.x
+
+            fn = self._functions.get(node.func.id)
+            if fn is None:
+                raise CompilerError(ERR_NO_DEF, node)
+
+            if len(node.args) != fn.argc:
+                raise CompilerError(ERR_ARGC_MISMATCH, node)
+
+            for arg in node.args:
+                val = self.as_value(arg)
+                self._ins.append(f'write {val} cell1 {REG_STACK}')
+                self._ins.append(f'op add {REG_STACK} {REG_STACK} 1')
+            # TODO  probably better to set the values straight into the relevant variables, rather than slowly piping through stack
+
+            # `REG_STACK` is not updated because it's immediately read by the function.
+            # If it was, the `op add` followed by the `op sub` would be redundant.
+            self._ins.append(f'write @counter cell1 {REG_STACK}')
+            self._ins.append(f'jump {fn.start} always')
+            return REG_RET
+            # TODO  probably better to pass the returned value via a dummy variable rather than piping through stack

```

Comment:
The returned value is not passed via the stack, it's stored in `REG_RET`. The `@counter` *is* in the stack though.